sql/src/main/java/io/crate/planner/node/dql/UnionPhase.java

```diff
@@ -0,0 +1,160 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.node.dql;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import io.crate.analyze.symbol.Symbols;
+import io.crate.planner.distribution.DistributionInfo;
+import io.crate.planner.distribution.UpstreamPhase;
+import io.crate.planner.node.ExecutionPhaseVisitor;
+import io.crate.planner.projection.Projection;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+import java.util.*;
+
+import static io.crate.planner.distribution.DistributionInfo.DEFAULT_SAME_NODE;
+
+public class UnionPhase extends AbstractProjectionsPhase implements UpstreamPhase {
+
+    public static final ExecutionPhaseFactory<UnionPhase> FACTORY = new ExecutionPhaseFactory<UnionPhase>() {
+        @Override
+        public UnionPhase create() {
+            return new UnionPhase();
+        }
+    };
+
+    private Collection<String> executionNodes;
+    private Collection<MergePhase> mergePhases;
+
+    public UnionPhase() {
+    }
+
+    public UnionPhase(UUID jobId,
+                      int executionNodeId,
+                      String name,
+                      List<Projection> projections,
+                      Collection<MergePhase> mergePhases,
+                      Collection<String> executionNodes) {
+        super(jobId, executionNodeId, name, projections);
+        assert !mergePhases.isEmpty();
+        if (projections.isEmpty()) {
+            outputTypes = Iterables.getLast(mergePhases).outputTypes();
+        } else {
+            outputTypes = Symbols.extractTypes(Iterables.getLast(projections).outputs());
+        }
+        this.mergePhases = mergePhases;
+        this.executionNodes = executionNodes;
+    }
+
+    @Override
+    public Type type() {
+        return Type.UNION;
+    }
+
+    @Override
+    public Collection<String> executionNodes() {
+        return executionNodes;
+    }
+
+    public Collection<MergePhase> mergePhases() {
+        return mergePhases;
+    }
+
+    @Override
+    public <C, R> R accept(ExecutionPhaseVisitor<C, R> visitor, C context) {
+        return visitor.visitUnionPhase(this, context);
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
```

Comment:
UnionPhase currently should never be streamed, throw an exception? (same at writeTo)